### PR TITLE
Upgrade to graphql-upload v8, dropping upload support for Node.js v6.

### LINKED ISSUE #1509: [Bug] Apollo server v2 restarts when receives invalid multipart-request operation
I was testing my api with postman and found that if incorrect multipart request operation is executed the server just restarts. 

Here's the example of request that I send to the server. 

```
operations:{"query":"query ($file:Upload){me(file: $file) { _id }}", "variables" : }
map:{"0": ["variables.file"] }
```

Notice that I'm passing file variable of type Upload to a me query (where it should not be), as a result I get an error `SyntaxError: Unexpected token } in JSON... ` And the app exits with code 1 
I don't think that it's a desirable behaviour in such situation

### LINKED ISSUE #1703: [File Upload] - flawed error handling for bad JSON
With "apollo-server-express": "^2.1.0".

I am testing my file upload API using a REST client (Insomnia). The Apollo Express Server delegates file upload handling to the Apollo Upload Server by Jayden Seric. This middleware handles multipart requests and requires a few HTTP headers to be set to work correctly. These headers ("operations" and "map") require JSON data. If I pass invalid JSON data to one of these headers, an unhandled exception is thrown _asynchronously_ in the Upload Server thus killing the whole server!

I believe this issue to have been fixed in later releases of the Upload Server. The latest Apollo Express uses apollo-upload-server v5. The apollo-upload-server code has been migrated to graphql-upload and the current version is 8. There is a small change in the API, which I believe to be non-essential. It would be nice to have the dependency updated! : ]

Working on Windows I find forking the apollo-server repo and testing this change a bit of a nuisance but if required I could try to do it.

### LINKED ISSUE #1854: Update to use grahpql-upload v8
> You can see in the graphql-upload (formerly apollo-upload-server) changelog that a lot of critical improvements have been made between v5 and v7. Some of the problems were very challenging to solve! It should be stable now and we have a lot of tests.
> 
> Everyone should migrate from apollo-upload-server v5 to graphql-upload v8+ ASAP.
> 

apollo-upload-server should be updated to use latest graphql-upload

### PR DESCRIPTION
**Pull request by abernix**
This PR builds upon #1764, and hopes to supersede it — though it stops short of the full deprecation of Node.js 6 that was offered in that solution.  Thank you to @lgaticaq for the work in getting that started since I merely needed to rebase this and manage the merge conflicts as the foundation for the rest of the work.  The remaining work was done for clarity and understanding to the users who will lose upload functionality on older Node.js versions because of this (necessary) upgrade.

Apollo Server utilizes @jaydenseric's (:wave:) [`graphql-upload` package](https://npm.im/graphql-upload) to implement out-of-the-box file upload functionality.  Changes and improvements to that package which fix potential crashes during file uploads **necessitate dropping support for file uploads in versions of the Node.js engine prior to v8.5.0** since the solution revolves around newer APIs which aren't available in older versions of Node.js.  There are great improvements and better use of modern APIs which improve upload functionality, so it's certainly worth the upgrade in other regards as well.

We'd normally try to abide by more traditional semantic versioning patterns but the solution here ultimately involved dropping support for an Node.js Long Term Support (LTS) version — v6.x — that is [still supported by the Node.js Foundation until April 2019](https://github.com/nodejs/Release#release-schedule).  This change brings a partial deprecation for those Node.js v6 users users if they're utilizing uploads.  It's worth noting that Node.js 8 did not become "LTS" until v8.9.0, and we aren't trying to support 8.0.0 through 8.5.0 since those are already _end-of-life'd_.

In an ideal world we could bump to 3.x but, realistically, there are other considerations for us when doing a major version bump — including maintaining backports and updating/forking documentation and educational content (much of which has just been published for 2.x!) — something we're not fully equipped to do at the moment.  And ultimately, there is just no solution for Node.js 6 who we still believe should update if they're using uploads.  Therefore, we've decided to `throw` if certain conditions are met on Node.js < 8.5 (for awareness!) and release this as a semver _minor_ version.

Why the `throw`? If we did not throw an error at server start-up, those affected may miss it.  Since file uploads are supported by default in Apollo Server 2.x, and there is an explicit dependency on `graphql-upload`, we must prevent users who are affected by this mid-major-release deprecation by being surprised by the sudden lack of upload support and give users a way to confirm that they don't need uploads since we can't detect it automatically. (I'm happy to ideas here, but I'm afraid we can't know with 100% certainty, which is why we've landed here.)  Therefore:

* Node.js 6 users who haven't explicitly disabled uploads already will need to disable uploads by setting `uploads: false`.
* Node.js 6 users who still need upload support and wish to update to the version of Apollo Server with this change will need to update to a newer Node.js LTS (e.g. v8.x, v10.x)

By `throw`-ing an error at server startup for affected users, we certainly are breaking a semantic versioning agreement for these users, but with good intention and a hopefully resounding reason. Luckily, with a relatively simple ergonomic (setting `uploads: false`) we allow those users who are NOT utilizing file uploads (as we believe is the case with a majority) to continue using their version of Node.js until it reaches the end of its supported lifetime (again, as per the [appropriate timelines](https://github.com/nodejs/Release#release-schedule)).

Apollo Server 2.x has attempted to maintain full compatibility with versions of Node.js which are still under LTS agreements with the Node.js Foundation.  While this continues to mostly be true, this is an exception which seems to still make sense.

While Node.js v6 is still under _Long Term Support_ from the Node.js Foundation, **we encourage _all users_ of Apollo Server to update to newer LTS versions of Node.js** prior to the "end-of-life" dates for their current server version.

For example, while Node.js 6.x is covered by Long Term Support until April 2019, there are substantial performance (e.g. [V8](https://v8.dev/) improvements) and language changes (e.g. "modern" ECMAScript support) offered by newer Node.js engines (e.g. 8.x, 10.x).  Switching to newer Long Term Support versions of Node.js comes with long-term benefits; Node.js 10.x LTS releases are covered by the Node.js Foundation through April 2021!

**We intend to drop support for Node.js 6.x in the next major version of Apollo Server.**

Lastly, this PR also adds documentation to help those affected — that documentation won't publish until it's merged (and the bit.ly link will 404 until it's published), but it will show on the Netlify preview of this PR for now.

Closes #1854 
Closes #1509 
Closes #1703
Closes #1764 (Supersedes)

### CODE REVIEW

**Comment by abernix** [comment]
Ok, Node.js 6 tests are mostly re-enabled now, thanks to https://github.com/apollographql/apollo-server/pull/2054/commits/2fd56f0c3e9ce677b06a09fe98b63dca09f65508.  Prior to that commit there were 300/563 failing tests but there are only a couple disabled Node.js 6 upload tests now after (the additional) https://github.com/apollographql/apollo-server/pull/2054/commits/cd6e575bb788036a9dfbdbc0aae14cc00286afa1 which excludes those.

Most happily, we're now actually testing Node.js 10 (implemented in https://github.com/apollographql/apollo-server/pull/2054/commits/705256e8790a4c7df84704bcc9e28f33b9b639d9) and file uploads – something we weren't doing before and I'm pretty happy to have moved on from that unfortunate exception.

File: packages/apollo-server-core/src/index.ts [hunk 1]
```diff
@@ -41,6 +41,18 @@ export const gql: (
...
-import { GraphQLUpload as UploadScalar } from '@apollographql/apollo-upload-server';
-export const GraphQLUpload = UploadScalar as GraphQLScalarType;
+export { default as processFileUploads } from './processFileUploads';
+
+// This is a conditional export intended to avoid traversing the
+// entire module tree of `graphql-upload`.  This only defined if the
+// version of Node.js is >= 8.5.0 since those are the only Node.js versions
+// which are supported by `graphql-upload@8`.  Since the source of
+// `graphql-upload` is not transpiled for older targets (in fact, it includes
+// experimental ECMAScript modules), this conditional export is necessary
```
Comment (jaydenseric) [inline packages/apollo-server-core/src/index.ts]: `graphql-upload` supports both CJS, *and* native ESM (if the consumers Node.js is running with the `--experimental-modules` flag). Both the CJS and ESM builds are transpilled, the only difference is the module format. But yes, there is a chance some new API or syntax might throw older Node.js if imported.

File: packages/apollo-server-core/src/ApolloServer.ts [hunk 2]
```diff
@@ -542,3 +552,32 @@ export class ApolloServerBase {
...
+      'The third-party `graphql-upload` package, which is used to implement',
+      'file uploads in Apollo Server 2.x, no longer supports Node.js LTS',
+      'versions prior to Node.js v8.5.0.',
+      '',
+      'If this server DOES NOT USE file uploads, it is necessary to add:',
+      '',
+      '  uploads: false,',
+      '',
+      'to the options for Apollo Server and re-deploy to disable file uploads',
+      'and continue using this version of Node.js.',
```
[the remark this answers is not in the document]
Comment (abernix) [reply packages/apollo-server-core/src/ApolloServer.ts]: Great point, @hwillson.  How about this take: https://github.com/apollographql/apollo-server/pull/2054/commits/fe2d5972c0f882763619c93c20de0d9efe0dda84

[the remark this answers is not in the document]
Comment (hwillson) [reply packages/apollo-server-core/src/ApolloServer.ts]: Looks great @abernix!

**Comment by abernix** [comment]
Merging into `release-vNEXT`.  I'll surface this in a `2.3.x-rc` soonish.

**Comment by abernix** [comment]
I've published this as `apollo-server@2.3.0-alpha.0`.

In order to gauge the success of the transition, it would be great to get feedback from any existing users who utilize uploads and are able to try it out!

**Comment by abernix** [comment]
The `alpha` releases didn't identify any problems so I've graduated this to the official `apollo-server-*@2.3.0` releases.

**Pull request merged**